lib/s3routes/routes/routeDELETE.ts

```diff
@@ -54,11 +54,11 @@ export default function routeDELETE(
         api.callApiMethod('objectDelete', request, response, log,
             (err, corsHeaders) => {
                 /*
-              * Since AWS expects a 204 regardless of the existence of
-              the object, the errors NoSuchKey and NoSuchVersion should not
-              * be sent back as a response.
-              */
-                if (err && !err.is.NoSuchKey && !err.is.NoSuchVersion) {
+                 * Since AWS expects a 204 regardless of the existence of
+                 * the object, the errors NoSuchKey and NoSuchVersion should not
+                 * be sent back as a response.
+                 */
+                if (err instanceof ArsenalError && !err.is?.NoSuchKey && !err.is?.NoSuchVersion) {
```

Comment:
* why would it not be an `ArsenalError` ? (i.e. are we not just "masking" an underlying issue: which should be fixed as well)
* if we receive an error, but not an ArsenalError : we change the behavior here... We probably don't want to include too much information indeed, but we should probably not send 204 either and report "some" error (maybe a generic 500?)

Reply:
> why would it not be an ArsenalError ? (i.e. are we not just "masking" an underlying issue: which should be fixed as well)

Because we return errors from the underlying modules, such as hdclient, in our APIs. These errors are not Arsenal errors. It's possible for example in the "DeleteObject" API if hdclient returns an error. It does not happen most of the time, but it can happen, and we should not make assumption in Arsenal for the error type: this was changed when typescript was implemented in arsenal, but our S3 implementation was not updated to always return ArsenalErrors. And it actually not only return ArsenalErrors.

> if we receive an error, but not an ArsenalError : we change the behavior here... We probably don't want to include too much information indeed, but we should probably not send 204 either and report some sort of error (maybe a generic 500?)

I am not sure to understand what you mean by "we change the behavior here": its the actual bug we are trying to fix. Reaching this line with a non-Arsenal error makes the process crash, because `is` is not a valid field from a normal error object.

Example:
```
 stderr F 2024-07-12T02:51:37.386Z: Uncaught exception: 
 stderr F TypeError: Cannot read properties of undefined (reading 'NoSuchKey')
 stderr F     at /usr/src/app/node_modules/arsenal/build/lib/s3routes/routes/routeDELETE.js:82:32
 stderr F     at /usr/src/app/lib/api/objectDelete.js:251:20
 stderr F     at /usr/src/app/node_modules/async/dist/async.js:421:16
 stderr F     at next (/usr/src/app/node_modules/async/dist/async.js:5302:29)
 stderr F     at /usr/src/app/node_modules/async/dist/async.js:906:16
 stderr F     at /usr/src/app/lib/api/objectDelete.js:211:41
 stderr F     at /usr/src/app/lib/services.js:372:36
 stderr F     at /usr/src/app/node_modules/hdclient/lib/hdcontroller.js:358:17
 stderr F     at /usr/src/app/node_modules/async/dist/async.js:421:16
```

Reply:
> I am not sure to understand what you mean by "we change the behavior here": it's the actual bug we are trying to fix

the previous behavior was to response w/o body (i.e. take the `if` branch) on any error except NoSuchKey / NoSuchVersion.

with the change, non-arsenal errors will not take this code path anymore : which is changing the behavior, not just fixing the crash...

the following would fix the crash without changing the behavior:

```suggestion
                if (err && !err.is?.NoSuchKey && !err.is?.NoSuchVersion) {
```

Reply:
AFAIU, we can now have non-arsenal errors from Vault, right?
and we may want to have 204 in such case?

Reply:
We always call `routesUtils.responseNoBody` in both cases, the only difference being the error code returned, but that does change your suggestion indeed.
However we cannot do that, if we do not test that the error is an arsenal error we get TS errors, so I adapted it a bit.

-> Updated, I also removed the last bump commit as per our discussion